ABI Compliance Checker Module

MankiratSingh1315 opened the pull request:
Introducing a new build farm module, ABICompCheck,
designed to perform Application Binary Interface (ABI) compliance checking
for PostgreSQL builds using libabigail tools abidiff and abidw.

The module helps detect unintended ABI changes in stable branches
by comparing the latest commit against the most recent tag on that branch or the most recent entry in `.abi-complaince-history` if present in the source.
This is crucial for maintaining compatibility
for the PostgreSQL ABI stability guidelines for minor releases.

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,259 @@
+
+	if (-f $last_commit_hash_file)
+	{
+		my $last_commit_hash=file_contents($last_commit_hash_file);
```
Perl style prefers spaces around operators:

```suggestion
		my $last_commit_hash = file_contents($last_commit_hash_file);
```

MankiratSingh1315 replied:
I discovered make file has a tidy command which automatically fixes all these - will use it from now on

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,259 @@
+
+	if (!defined $last_commit_hash || $last_commit_hash eq '')
+	{
+		print time_str(),
+		  "ABICompCheck: No previous commit hash found, this runs.\n"
+		  if $verbose;
+		return;
```
Since your return, it seems like it doesn't run. Is "this runs" on line 192 accurate?

MankiratSingh1315 replied:
that was added so that the further logic meant for a group of backlog commits wont run
could put that in else part as well and remove the return

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,259 @@
+		return;
+	}
+
+	my @commits = `git rev-list --reverse $last_commit_hash..HEAD`;
```
I think you should use the PGBuild::SCM module for all interactions with the repository. If it doesn't provide the function you need, add it.

MankiratSingh1315 replied:
Okay will look into this now
but the scm module seems to be whole new story - it was the reason to cause delay for me since its too much complicated 💀

theory replied:
I think you just need to look for the features you need and use them. But maybe there isn't one to check out a specific commit?

MankiratSingh1315 replied:
Yes, Have to create a function for them all

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,258 @@
+		$abi_compare_root = "$buildroot/abicheck";
+	}
+
+	mkdir $abi_compare_root unless -d $abi_compare_root;
```
Yes the run_build.pl uses this only for creating dirs so used it

MankiratSingh1315 commented:
will be working on the minor bugs like the abidw flags and the postgres cflags take them from config,
last_commit_hash_file thing is still not working, will fix it as well

![image](https://github.com/user-attachments/assets/24072802-a349-4f94-88b6-c0c766f79e24)
this is the output directory structure as of now

MankiratSingh1315 commented:
Output Logs as of now
 
```
Tue Jun 17 02:59:13 2025: buildfarm run for df4c2566:HEAD starting
df4c2566:HEAD          [02:59:14] checking out source ...
df4c2566:HEAD          [02:59:17] checking if build run needed ...
df4c2566:HEAD          [02:59:17] checking if run needed by PGBuild::Modules::ABICompCheck
df4c2566:HEAD          [02:59:17] ABICompCheck: Stored PostgreSQL commit hash file ('/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/HEAD/last_pg_commit_hash.txt') not found. Run needed.
df4c2566:HEAD          [02:59:17] copying source to pgsql.build ...
df4c2566:HEAD          [02:59:23] building PGBuild::Modules::ABICompCheck
running git clone -q  "/home/pgbuildfarm/client-code/buildroot/pgmirror.git" "pgsql"
df4c2566:HEAD          [02:59:26] ABICompCheck: Cloned PostgreSQL source to /home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/pgsql
df4c2566:HEAD          [02:59:27] ABICompCheck: Checked out branch 'HEAD' in /home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/pgsql
df4c2566:HEAD          [02:59:27] ABICompCheck: No previous commit hash found, comparing last 2 commits only.
Two commit hashes: ee685c9baf987984dbd2194d74576422996d95f4 f24fdf985561e7166d7d54459b764daae8efda44
df4c2566:HEAD          [02:59:27] Processing commits list in PGBuild::Modules::ABICompCheck
df4c2566:HEAD          [02:59:27] Checking out commit ee685c9baf987984dbd2194d74576422996d95f4 in PGBuild::Modules::ABICompCheck
df4c2566:HEAD          [02:59:27] Configuring, making, and installing for commit ee685c9baf987984dbd2194d74576422996d95f4 in PGBuild::Modules::ABICompCheck
df4c2566:HEAD          [02:59:27] Executing: configure
df4c2566:HEAD          [02:59:46] Successfully executed configure
df4c2566:HEAD          [02:59:46] Configuration complete for commit ee685c9baf987984dbd2194d74576422996d95f4.
df4c2566:HEAD          [02:59:46] Executing: make
df4c2566:HEAD          [03:05:52] Successfully executed make
df4c2566:HEAD          [03:05:52] Executing: makeinstall
df4c2566:HEAD          [03:05:55] Successfully executed makeinstall
df4c2566:HEAD          [03:05:55] Generating ABIDW XML for commit ee685c9baf987984dbd2194d74576422996d95f4 in PGBuild::Modules::ABICompCheck
df4c2566:HEAD          [03:05:55] Executing: abidw --out-file "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/ee685c9baf987984dbd2194d74576422996d95f4/postgres.abi" "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/install/bin/postgres" --drop-undefined-syms --no-architecture --no-comp-dir-path --no-elf-needed --no-show-locs --type-id-style hash
df4c2566:HEAD          [03:06:10] Successfully generated ABI XML for postgres to /home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/ee685c9baf987984dbd2194d74576422996d95f4/postgres.abi
df4c2566:HEAD          [03:06:10] Executing: abidw --out-file "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/ee685c9baf987984dbd2194d74576422996d95f4/ecpg.abi" "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/install/bin/ecpg" --drop-undefined-syms --no-architecture --no-comp-dir-path --no-elf-needed --no-show-locs --type-id-style hash
df4c2566:HEAD          [03:06:11] Successfully generated ABI XML for ecpg to /home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/ee685c9baf987984dbd2194d74576422996d95f4/ecpg.abi
df4c2566:HEAD          [03:06:11] Executing: abidw --out-file "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/ee685c9baf987984dbd2194d74576422996d95f4/libpq.so.abi" "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/install/lib/libpq.so" --drop-undefined-syms --no-architecture --no-comp-dir-path --no-elf-needed --no-show-locs --type-id-style hash
df4c2566:HEAD          [03:06:11] Successfully generated ABI XML for libpq.so to /home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/ee685c9baf987984dbd2194d74576422996d95f4/libpq.so.abi
df4c2566:HEAD          [03:06:11] Checking out commit f24fdf985561e7166d7d54459b764daae8efda44 in PGBuild::Modules::ABICompCheck
df4c2566:HEAD          [03:06:11] Configuring, making, and installing for commit f24fdf985561e7166d7d54459b764daae8efda44 in PGBuild::Modules::ABICompCheck
df4c2566:HEAD          [03:06:11] Executing: configure
df4c2566:HEAD          [03:06:29] Successfully executed configure
df4c2566:HEAD          [03:06:29] Configuration complete for commit f24fdf985561e7166d7d54459b764daae8efda44.
df4c2566:HEAD          [03:06:29] Executing: make
df4c2566:HEAD          [03:06:37] Successfully executed make
df4c2566:HEAD          [03:06:37] Executing: makeinstall
df4c2566:HEAD          [03:06:41] Successfully executed makeinstall
df4c2566:HEAD          [03:06:41] Generating ABIDW XML for commit f24fdf985561e7166d7d54459b764daae8efda44 in PGBuild::Modules::ABICompCheck
df4c2566:HEAD          [03:06:41] Executing: abidw --out-file "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/f24fdf985561e7166d7d54459b764daae8efda44/postgres.abi" "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/install/bin/postgres" --drop-undefined-syms --no-architecture --no-comp-dir-path --no-elf-needed --no-show-locs --type-id-style hash
df4c2566:HEAD          [03:06:55] Successfully generated ABI XML for postgres to /home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/f24fdf985561e7166d7d54459b764daae8efda44/postgres.abi
df4c2566:HEAD          [03:06:55] Executing: abidw --out-file "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/f24fdf985561e7166d7d54459b764daae8efda44/ecpg.abi" "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/install/bin/ecpg" --drop-undefined-syms --no-architecture --no-comp-dir-path --no-elf-needed --no-show-locs --type-id-style hash
df4c2566:HEAD          [03:06:55] Successfully generated ABI XML for ecpg to /home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/f24fdf985561e7166d7d54459b764daae8efda44/ecpg.abi
df4c2566:HEAD          [03:06:55] Executing: abidw --out-file "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/f24fdf985561e7166d7d54459b764daae8efda44/libpq.so.abi" "/home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/install/lib/libpq.so" --drop-undefined-syms --no-architecture --no-comp-dir-path --no-elf-needed --no-show-locs --type-id-style hash
df4c2566:HEAD          [03:06:55] Successfully generated ABI XML for libpq.so to /home/pgbuildfarm/client-code/buildroot/abicheck.df4c2566/xmls/f24fdf985561e7166d7d54459b764daae8efda44/libpq.so.abi
df4c2566:HEAD          [03:06:55] OK
Branch: HEAD
All stages succeeded
```

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -166,88 +143,307 @@ sub need_run
+
+	open my $fh, '>', $log_file
+	  or die "Cannot open log file $log_file for writing: $!";
+	print $fh "$_\n" for @output;
```
Where is `run_log` defined? Depending on how it generates the output, you might need to tweak things to encode `@output` before you write it out.

Ah, I see it in Utils.pm. Looks like it doesn't set the encoding, to it will default to Latin-1. Your use of `>` on line 182 is equivalent. Probably fine for now.

theory replied:
I am wondering, though, whether the commands sent to `run_log` could simply write directly to the log file using `>$log_file 2>&`. Maybe that would fail on Windows, tho. Seems silly to load all the output into memory and then write it out, rather than just write it out…

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -166,88 +143,307 @@ sub need_run
-			or die "Failed to checkout $commit";
+		my @checkout_log = run_log(qq{git checkout -q "$branch"});
+		print_logs(\@checkout_log);
+		$status = $? >> 8;
```
I think this needs to go before `print_logs()`, no? Might `print_logs()` clear `$?`?

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+	my $pgsql = shift;        # postgres build dir
+
+	# We are only testing HEAD and stable branches, so ignore all others.
+	return if $branch !~ /^(?:HEAD|REL_?\d+(?:_\d+)?_STABLE)$/;
```
I think this can be simplified:

```suggestion
	return if $branch ne 'HEAD' && $branch !~ /_STABLE$/;
```

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+		$abi_compare_root = "$buildroot/abicheck";
+	}
+
+	my %binaries_rel_path = $conf->{binaries_rel_path}
+	  || (
+		'postgres' => 'bin/postgres',
+		'ecpg' => 'bin/ecpg',
+		'libpq.so' => 'lib/libpq.so',
+	  );
```
You can't assign a scalar to a list value (in this case a hash). Since `$conf->{binaries_rel_path}` can only be a reference, you should just work with references:

```suggestion
	my $binaries_rel_path = $conf->{binaries_rel_path} || {
		'postgres' => 'bin/postgres',
		'ecpg' => 'bin/ecpg',
		'libpq.so' => 'lib/libpq.so',
  	};
```

MankiratSingh1315 replied:
Oh yes, my bad

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+		'libpq.so' => 'lib/libpq.so',
+	  );
+
+	my @abidw_flags_list = $conf->{abidw_flags_list}
+	  || (
+		'--drop-undefined-syms', '--no-architecture', '--no-comp-dir-path',
+		'--no-elf-needed', '--no-show-locs', '--type-id-style',
+		'hash',
+	  );
```
Same deal: work with referees:

```suggestion
	my $abidw_flags_list = $conf->{abidw_flags_list} || [
		'--drop-undefined-syms', '--no-architecture', '--no-comp-dir-path',
		'--no-elf-needed', '--no-show-locs', '--type-id-style',
		'hash',
	];
```

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+	}
+	else
+	{
+		print "running ", qq{git clone -q  "$mirror_base" "$clone_name"}, "\n";
+		@clonelog = run_log(qq{git clone -q "$mirror_base" "$git_repo_path"});
```
I'm wondering why you're cloning here. Can you not work with branches directly?

MankiratSingh1315 replied:
As far as I have understood the code, they are not copying the .git folder directly, instead just copying the source of the latest commit for running the tests on limited branches which is good option for the testing purpose but it don't seems to work for iteraritng through the commits so I just chose to clone from the mirror.

https://github.com/PGBuildFarm/client-code/blob/b99f853c799ebbecc15962ad31d8a136cbe88a6b/PGBuild/SCM.pm#L549

theory replied:
Got it. I suggest using `--depth 1 --branch $branch` to minimize I/O.

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+		}
+	}
+
+	my @checkout_log =
+	  run_log(qq{git -C "$git_repo_path" checkout -q "$branch"});
+	$status = $? >> 8;
+	print_logs(\@checkout_log);
+	if ($status)
+	{
+		die "git checkout failed with status $status";
+	}
```
The SCM Module has a `checkout` method; this should work, no?

```suggestion
	print_logs($self->{scm}->checkout($branch));
```

MankiratSingh1315 replied:
Similar to the previous comment, they are not working directly with git repos, instead copying the source of branches from the mirror - I tried to workaround this checkout method a few days back, but I'm either I am unable to figure out, or maybe it actually doesn't work in case of this project

MankiratSingh1315 replied:
I won't be able to loop through commits in that case, if I am not wrong?

theory reviewed (changes requested):
Finally made a careful read and made some suggestions. Seems like the basic functionality is fine, though we may want to do some refactoring later. In fact, I think it would be nice to factor a bunch of the `configure` and `make` code from `run_build.pl` into a module that could then be called from here, so that the ABI Check gets the benefit of the full configuration (e.g., builds with libxml, LDAP, ICU, Perl, Python, etc.).

After a run, I see this directory layout:

```tree
.
|-- diffs
|   |-- ecpg-fe07100e82b096d3c848cace790d4b4daf0c4131-9ca30a0b04d751c58d4efa0a2b3073cb285b1bd2.log
|   `-- postgres-fe07100e82b096d3c848cace790d4b4daf0c4131-9ca30a0b04d751c58d4efa0a2b3073cb285b1bd2.log
|-- githead.log
|-- logs
|   |-- 9ca30a0b04d751c58d4efa0a2b3073cb285b1bd2
|   |   |-- configure.log
|   |   |-- make.log
|   |   `-- makeinstall.log
|   `-- fe07100e82b096d3c848cace790d4b4daf0c4131
|       |-- configure.log
|       |-- make.log
|       `-- makeinstall.log
`-- xmls
    |-- 9ca30a0b04d751c58d4efa0a2b3073cb285b1bd2
    |   |-- ecpg.abi
    |   |-- libpq.so.abi
    |   `-- postgres.abi
    `-- fe07100e82b096d3c848cace790d4b4daf0c4131
        |-- ecpg.abi
        |-- libpq.so.abi
        `-- postgres.abi
```

Which of those files do you plan to upload to the server?

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+	  "Cannot change to PostgreSQL source directory: $abi_compare_root/$clone_name for commit $commit_hash";
+
+	_log_command_output($self,
+		qq{./configure CFLAGS="-Og -g" --prefix=$abi_compare_root/install/},
```
```suggestion
		qq{./configure CFLAGS="-Og -g" --without-icu --prefix=$abi_compare_root/install/},
```

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+		qq{./configure CFLAGS="-Og -g" --prefix=$abi_compare_root/install/},
+		$log_dir, 'configure');
+
+	_log_command_output($self, qq{make}, $log_dir, 'make');
+
+	_log_command_output($self, q{make install}, $log_dir, 'makeinstall');
```
```suggestion
	my $make = $self->{bfconf}{make} . ($self->{bfconf}{make_jobs} ? " -j $self->{bfconf}{make_jobs}": '');
	_log_command_output($self, $make, $log_dir, 'make');

	_log_command_output($self, qq{$make install}, $log_dir, 'makeinstall');
```

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+# Package Namespace is hardcoded. Modules must live in
+# PGBuild::Modules
+
```
```suggestion

# apt-get install -y build-essential abigail-tools libio-all-lwp-perl git ccache bison flex libreadline-dev zlib1g-dev
```

theory replied:
Yeah, these were kind of notes I had for myself trying it out on Debian. I currently have

``` perl
=pod
apt-get install -y build-essential abigail-tools libio-all-lwp-perl git ccache \
   bison flex libreadline-dev zlib1g-dev apt-get install -y libperl-dev\
   libipc-run-perl uuid-dev
=cut
```

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,641 @@
+				name => $target_name,
+				input_path => $input_path,
+				output_file => $output_file,
+			  };
```
Just for a better readability and I was initially writing it, I have remove the duplicate code now 😅 🫡

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,720 @@
+	my $abi_compare_loc = "$self->{abi_compare_root}/$self->{pgbranch}";
+
+	emit "checking if run needed";
+	return;
```
I thought about this several times, but...there's no particular condition on why "run is not needed". Like if the `ABICompCheck` module is mentioned in config then just let the module work, no?
Conditions like run needed based on if there's some new commit or correctly compiled from source are already handled by the `run_build.pl` script. In case the xmls for latest tag not found or something similar, that all is handled in install function.


Maybe I should remove the usage of this `need_run` method from hooks in first place, just left it there because other module like `TextUpgradeXversion.pm` was also having it without any condition😅

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,760 @@
+		run_log(qq{git -C ./pgsql checkout $comparison_ref});
+
+		# got this git save peice of code from PGBuild::SCM::Git::copy_source
+		move "./pgsql/.git", "./git-save";
```
From what I’ve seen, each run’s working directory is `<buildroot>/<branch_name>`, so the `.git` folder belongs to the pgsql source code of a particular branch and is used only by a particular run.

The SCM module was doing the same thing in checkout function. But yes, it could fail in case another animal is running on the same machine. I'll look into this 👍🏻

theory replied:
You may be right. If it doesn't fail don't worry about it.

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,760 @@
+		}
+	}
+
+	if (-d "./inst")
```
Install files/binaries for the most recent commit

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,760 @@
+		# Check if XML files for latest tag exist, generate if missing
+		my $tag_inst_dir = "$abi_compare_loc/$comparison_ref/inst";
+		my $tag_xml_dir = "$abi_compare_loc/$comparison_ref/xmls";
+		if (-d $tag_inst_dir)
```
Great. Just needs comments/docs to explain that, I think.

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,760 @@
+		{
+			$diff_found = 1;
+			my $log_file = "$log_dir/$key-$latest_tag.log";
+			emit "ABI difference for $key: one file is missing (tag: $tag_file, branch: $branch_file). Comparison skipped.";
```
You are right!
AI overtakes my mind sometimes, sorry ;-;

theory replied:
Please lean less on chatbots and more on your own discernment. Pavlo and I are always happy to help answer questions or propose code! You'll build up more expertise doing it yourself and thinking through all the branches as you work.

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,760 @@
+	{
+		my @saveout;
+		push(@saveout, $log->log_string);
+		writelog("abi-comp-check-$current_branch", \@saveout);
```
Yess
Line 678, i.e. creating the log object is required step to have the writelog function working.

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,760 @@
+sub cleanup
+{
+	my $self = shift;
+	if (!$keepall) {
```
`Options.pm` - Line 45

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,760 @@
+	my $abi_compare_root = $self->{abi_compare_root};
+	my $pgbranch = $self->{pgbranch};
+
+	emit "Comparing ABI between latest tag $latest_tag and it's latest commit";
```
```suggestion
	emit "Comparing $pgbranch ABI between $latest_tag and its latest commit $commit_sha";
```

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,877 @@
+=back
+
+=head2 CONFIGURATION OPTIONS
+The module supports the following configuration options:
```
Oh, yes
forgot about that 😅

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,720 @@
+	# We are only testing stable branches, so ignore all others.
+	return if $branch !~ /_STABLE$/;
+	return if $conf->{using_msvc};
+	return if $conf->{scm} ne 'git';
```
I'm not sure about that, like what if someone's using MacOS(maybe that will also be considered under linux?) and what if someone using WSL on windows?

Also, I do think meson has cross platform support...

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,983 @@
+
+		# Build the tag: configure, make, install
+		my @configure_log = $self->configure($abi_compare_loc, $comparison_ref);
+		my @make_log = $self->make($abi_compare_loc, $comparison_ref);
```
I think that's because they're written elsewhere. It doesn't appear they're used here.

MankiratSingh1315 replied:
> Although it doesn't look like you use it at all? Why return the logs if you don't use them?

Eh ok, I remember I had plans initially to push these logs directly to `@saveout` but that cased the issue that I can't get that expand to view log on the build farm server UI and it is considered as plain text on it
So had to use the add_log function for it which resulted in me forgetting to remove these variables🫠

MankiratSingh1315 committed:
Introducing a new build farm module, ABICompCheck,
designed to perform Application Binary Interface (ABI) compliance checking
for PostgreSQL builds using libabigail tools abidiff and abidw.

The module helps detect unintended ABI changes in stable branches
by comparing the latest commit against the most recent tag on that branch.
This is crucial for maintaining compatibility
for the PostgreSQL ABI stability guidelines for minor releases.

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,1010 @@
+
+	emit "======== setup output ===========\n", @confout if ($verbose > 1);
+
+	# writelog($latest_tag . 'configure', \@confout);
+
+	# if ($status)
+	# {
+	# 	send_result('Configure', $status, \@confout);
+	# }
+
+	# return @confout;
```
What's the reason to keep this commented code?

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,1010 @@
+	print $fh @confout;
+	close $fh;
+
+	# if ($status)
+	# {
+	# 	send_result('Configure', $status, \@confout);
+	# }
+
+	# return \@confout;
```
Again it seems to me some of this commented code can be removed, yes?

theory commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,1010 @@
+		chomp $tag_commit;
+
+		my $is_ancestor = system(qq{git -C ./pgsql merge-base --is-ancestor $tag_commit $comparison_ref 2>/dev/null});
+		if ($is_ancestor != 0) {
```
```suggestion
		if ($is_ancestor) {
```

theory replied:
Yeah, in Perl, anything other than `0`, `''`, and `undef` is true.

theory reviewed (approved):
Just a few minor nits. Looks great thank you!

theory committed:
Some minor Pod reformatting

Plus one little punctuation fix.

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,1010 @@
+	mkdir $abi_compare_root
+	  unless -d $abi_compare_root;
+
+	# we need to segregate from-source builds so they don't corrupt
+	# non-from-source saves
+
+	# my $fs_abi_compare_root = "$buildroot/fs-upgrade.$animal";
+
+	# mkdir $fs_abi_compare_root
+	#   if $from_source && !-d $fs_abi_compare_root;
```
No idea about how to actually use this option, found this part in some other module

Some mention for from-source is in the bf howto: https://wiki.postgresql.org/wiki/PostgreSQL_Buildfarm_Howto#All_the_run_build.pl_command_line_options

MankiratSingh1315 replied:
Okayy

MankiratSingh1315 committed:
minor fixes

MankiratSingh1315 commented on PGBuild/Modules/ABICompCheck.pm:
```diff
@@ -0,0 +1,986 @@
+
+  {
+    'postgres' => 'bin/postgres',
+    'ecpg' => 'bin/ecpg',
```
Ok, so I noticed that the ABICompCheck module isn’t currently handling the installation of contrib and test modules, unlike what’s done in the [run_build.pl script](https://github.com/PGBuildFarm/client-code/blob/1f7cfa7a0a4c51ad954a7334b37d650fb26f56e8/run_build.pl#L1458-L1556). Because of this, a lot of `.so` files are missing from the baseline tag install. Could you please confirm whether these `.so` files are actually required for ABI compliance checking?

I never really knew about contrib and test modules before, but after reading up on them, it seems like the `.so` files for contrib would be a good to have in the ABI compliance check report, right?

petere replied:
> Ok, so I noticed that the ABICompCheck module isn’t currently handling the installation of contrib and test modules, unlike what’s done in the [run_build.pl script](https://github.com/PGBuildFarm/client-code/blob/1f7cfa7a0a4c51ad954a7334b37d650fb26f56e8/run_build.pl#L1458-L1556). Because of this, a lot of `.so` files are missing from the baseline tag install. Could you please confirm whether these `.so` files are actually required for ABI compliance checking?

They are not.

MankiratSingh1315 replied:
Done with the changes!

Really sorry for the delay, I got stuck with some of my college assignments 🙇🏻

petere replied:
You only need the compare

bin/postgres
lib/libecpg.so
lib/libecpg_compat.so
lib/libpgtypes.so
lib/libpq.so

The rest is not relevant.

theory replied:
I copied this updated version to baza.

MankiratSingh1315 committed:
Refactor ABI comparison logic to dynamically determine binaries and remove redundant logging

MankiratSingh1315 committed:
documentation update

MankiratSingh1315 committed:
Improved logic for polulating binaries_rel_path and perltidy

MankiratSingh1315 committed:
put list of binaries compared in abi comp check results

MankiratSingh1315 committed:
Refine binary path construction in ABICompCheck to explicitly scan .so files in inst/lib

MankiratSingh1315 committed:
Add support for .abi-compliance-history file for having custom baselines AND update docs

MankiratSingh1315 committed:
Improve code readability and perltidy

MankiratSingh1315 committed:
only work on presence of .abi-compliance-history file, update docs, update variable name for better code understanding

MankiratSingh1315 committed:
Move `.abi-compliance-history` required check to `installcheck`

When `rm_worktree` is enabled in animal config, the source directory is cleaned after the
run, causing `.abi-compliance-history` to be missing in subsequent runs as setup runs before source checkout.

adunstan commented:
@theory yes, thanks for the reminder
